### DISCUSSION TITLE
How to write query to get Call Graph in python?

### DISCUSSION THREAD

**Discussion by vzcrab** [github/codeql#8063] [Q&A]
And I want to use `@kind graph` to draw a graph. Although I find the  `FunctionInvocation` in the CallGraph module, it's hard to get what I want.

**Comment by smowton** [comment]
How about

```
/**
 * @kind graph
 */

import semmle.python.pointsto.CallGraph

query predicate edges(CallNode caller, CallNode callee) {
  exists(FunctionInvocation callerFI, FunctionInvocation calleeFI |
    callerFI.getCall() = caller and
    calleeFI.getCall() = callee and
    callerFI.getACallee() = calleeFI and
    calleeFI = callerFI.getACallee()
  )
}
```

NB. I'm unfamiliar with that callgraph library, so Python team members might be able to recommend a better way!

**Reply by vzcrab** [reply]
Thanks, I probably get know how to use query predicate for the edges. The `FunctionInvocation` has a comment that "This class represents a static approximation to the dynamic call-graph. A FunctionInvocation represents all calls made to a function for a given context." 

So the `FunctionInvocation` seems to supply a small set, and its handling is a little strange. I don't know how to control it.

**Reply by RasmusWL** [reply]
Hi @SINC-G, I work on the Python analysis, and if I had to show the calls we currently resolve, I would use the code below. It does not provide a `@kind graph` solution, but you can adapt it if you want. I would not use `FunctionInvocation` directly.

```codeql
private import python

from Value val, CallNode call
where call = val.getACall()
select val, call
```

**Reply by vzcrab** [reply]
Very thanks @RasmusWL. I'm sorry to reply so late.

Thank you for giving me a good way. Now I have an idea about how to achieve it.

**Reply by vzcrab** [reply]
@RasmusWL I'm sorry border you again. I find that Codeql Python analysis did not get the complete calling relationship. 

```QL
import python

from Value val, CallNode call
where
  val.getACall() = call
select call.getScope(), val
```
Using the above query, the method call of the instance in the query result can only get the instance method call that appears first in the source code file, and other method calls of the instance cannot be analyzed.

```python
class VulnDetection(object):
    def __init__(self, file, output=settings.DATA_PATH) -> None:
        self.file = file
        self.output = output
        self.vmatch = MATCH(settings.BASE_PATH, self.file, self.output) 

    def decompile(self) -> bool:
        return self.vmatch.decompile()

    def scan_url(self):
        result = self.vmatch.match_url()
        return result
```
As in the above example, it can analyze the call `self.vmatch.decompile()` to `MATCH.decompile`, but it is not analyzed the call `self.vmatch.match_url ()`.

And if you only use `CallNode`, you can query the call record of `self.vmatch.match_url ()`.

I hope you can give me some tips to resolve this problem for getting all caller and callee.

**Reply by RasmusWL** [reply]
Hi @SINC-G. The call graph you're able to get with the above code snippet is using our points-to analysis. We are aware there are problems with resolving all calls correctly, and I'm currently working on rewriting it.

If you're able to share a more complete example of your problem, I can include this in my test-suite :blush:

Just to be clear, the ql code I above _is_ really the way to do it right now.

**Reply by vzcrab** [reply]
Thanks for your reply @RasmusWL. I'm glad to hear that.

The example code is in this repository. https://github.com/SINC-G/Melody

I hope this will help you.

**Reply by andrescodas** [reply]
Hi!... I'm working on this use case.  I'm wondering if you can point me to an example where more information about the callee is displayed, like where it is defined, its module, qualifiedName, etc...  

My end goals it to be able to tell whether a call within a certain scope, is targeting a callee that is defined in another module or file

Thanks so much!

**Reply by RasmusWL** [reply]
Hi @andrescodas, for something like calls to functions/classes in different modules, you could use something like this:

```codeql
import python

from Value val, CallNode call, Scope scope
where
  val.getACall() = call and
  (
    scope = val.(CallableValue).getScope()
    or
    scope = val.(ClassValue).getScope()
  ) and
  not call.getEnclosingModule() = scope.getEnclosingModule()
select call, val, scope
```

if you want to restrict it to only be for calls and functions/classes in user-written code, you could do this instead:

```codeql
import python

from Value val, CallNode call, Scope scope
where
  val.getACall() = call and
  (
    scope = val.(CallableValue).getScope()
    or
    scope = val.(ClassValue).getScope()
  ) and
  not call.getEnclosingModule() = scope.getEnclosingModule() and
  exists(scope.getLocation().getFile().getRelativePath()) and
  exists(call.getLocation().getFile().getRelativePath())
select call, val, scope
```

happy coding :+1:

**Reply by andrescodas** [reply]
Hi @RasmusWL ,

Thanks for your help so far. I have a new related question.

I have been using this query to build a Call Graph:

```ql
import python 

from Value val,
     CallNode call,
     string call_location,
     string caller_scope_location,
     string callee_definition_location
where
  call = val.getACall() and
  (
    callee_definition_location = val.(CallableValue).getScope().getLocation().toString()
    or
    callee_definition_location = val.(ClassValue).getScope().getLocation().toString()
  ) and
  caller_scope_location = call.getScope().getLocation().toString() and
  call_location = call.getLocation().toString() 
select call_location, caller_scope_location, callee_definition_location
```

However, this didnt work as expected for the call to `foreach` [here](https://github.com/0101/pipetools/blob/9fa90986ce044e3f40f78a96480e2b5e5eb2b9da/test_pipetools/test_decorators.py#L10)

Particularly, I was expecting to see in the output this tuple:


  [
    ["../test_pipetools/test_decorators.py:10"](https://github.com/0101/pipetools/blob/9fa90986ce044e3f40f78a96480e2b5e5eb2b9da/test_pipetools/test_decorators.py#L10),
    ["../test_pipetools/test_decorators.py:9"](https://github.com/0101/pipetools/blob/9fa90986ce044e3f40f78a96480e2b5e5eb2b9da/test_pipetools/test_decorators.py#L9),
    ["../pipetools/utils.py:24"](https://github.com/0101/pipetools/blob/9fa90986ce044e3f40f78a96480e2b5e5eb2b9da/pipetools/utils.py#L24)
  ],


Can you help me improve my query so I catch this case too?

**Reply by RasmusWL** [reply]
> Can you help me improve my query so I catch this case too?

sounds like this could just be a case of the call not being resolved properly
